# **Travel Planner RAG System for Nepal (...Continued)**

# RAG Project - Notebook 2: Retrieval, LLM Inference & Evaluation


## Summary of Notebook 1

In the previous notebook, we:

- Ingested 9 curated `.txt` files covering Nepalese travel topics (food, treks, visas, etc.)
- Applied **file-specific chunking logic** to split long texts into semantically meaningful pieces
- Used the `all-MiniLM-L6-v2` model to **vectorize** the 104 final chunks
- Built a **FAISS index** using L2 similarity for fast retrieval
- Saved the index (`nepal_rag_index.faiss`) and chunks (`nepal_chunks.pkl`) for use in this notebook


This notebook focuses on running inference using 5 different LLMs and comparing their answers to 10 domain-specific travel questions.


## Table of Contents

1. Load FAISS Index and Chunks
2. Load Embedding Model for Queries
3. Load 5 LLMs (FLAN, TinyLLaMA, etc.)
4. Define RAG Prompt Generator per Model
5. Input: 10 Evaluation Questions
6. RAG Loop: Run Inference with All Models
7. Side-by-Side Comparison Table
8. Result Analysis & Observations

### **RAG Architecture Summary**

```plaintext
[.txt files]
    ↓
[Custom Chunking (file-specific)]
    ↓
[MiniLM Embedding: 384-d vectors]
    ↓
[FAISS Index: L2 Similarity]
    ↓
[User Query → MiniLM Vector]
    ↓
[Top k=4 Matching Chunks]
    ↓
[RAG Prompt Generator (per model)]
    ↓
[FLAN-Base / TinyLLaMA / FLAN Large]
    ↓
[Answer Generation]


# Installing libraries and dependencies

In [1]:
!pip install -q \transformers==4.36.2 \ faiss-cpu

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 126.8/126.8 kB 4.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.2/8.2 MB 99.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 31.3/31.3 MB 69.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 71.6 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
sentence-transformers 3.4.1 requires transformers<5.0.0,>=4.41.0, but you have transformers 4.36.2 which is incompatible.


In [2]:
from transformers import AutoTokenizer, AutoModel, AutoModelForCausalLM, AutoModelForSeq2SeqLM, pipeline, T5ForConditionalGeneration
import torch
import faiss
import pickle
import pandas as pd
from IPython.display import display

/usr/local/lib/python3.11/dist-packages/transformers/utils/generic.py:441: FutureWarning: `torch.utils._pytree._register_pytree_node` is deprecated. Please use `torch.utils._pytree.register_pytree_node` instead.
  _torch_pytree._register_pytree_node(
/usr/local/lib/python3.11/dist-packages/transformers/utils/generic.py:309: FutureWarning: `torch.utils._pytree._register_pytree_node` is deprecated. Please use `torch.utils._pytree.register_pytree_node` instead.
  _torch_pytree._register_pytree_node(


In [3]:
print("CUDA available?", torch.cuda.is_available())

CUDA available? True


# 1. Load FAISS index + chunks

In [4]:
index = faiss.read_index("nepal_rag_index.faiss")

with open("nepal_chunks.pkl", "rb") as f:
    chunks = pickle.load(f)

print(f"Loaded FAISS index and {len(chunks)} text chunks.")

Loaded FAISS index and 104 text chunks.


# 2. Load query embedder (TransformersEmbedder)

In [5]:
class TransformersEmbedder:
    def __init__(self, model_id="sentence-transformers/all-MiniLM-L6-v2"):
        self.tokenizer = AutoTokenizer.from_pretrained(model_id)
        self.model = AutoModel.from_pretrained(model_id).eval().to("cuda" if torch.cuda.is_available() else "cpu")

    def encode(self, texts):
        if isinstance(texts, str):
            texts = [texts]
        inputs = self.tokenizer(texts, padding=True, truncation=True, return_tensors="pt").to(self.model.device)
        with torch.no_grad():
            outputs = self.model(**inputs)
        return outputs.last_hidden_state[:, 0, :].cpu().numpy()

In [6]:
embedder = TransformersEmbedder()

/usr/local/lib/python3.11/dist-packages/huggingface_hub/file_download.py:896: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

## RAG Chunk Retrieval Test

In [7]:
def search_faiss_index(query, embedder, index, chunks, k=4):
    query_embedding = embedder.encode(query).astype("float32")
    _, I = index.search(query_embedding, k)
    return [chunks[i] for i in I[0]]

In [8]:
query = "What are the must-try foods for tourists visiting Kathmandu?"
top_chunks = search_faiss_index(query, embedder, index, chunks, k=4)

for i, chunk in enumerate(top_chunks, 1):
    print(f"--- Chunk {i} ---\n{chunk}\n")

--- Chunk 1 ---
6. Chatamari, or 'Nepali Pizza'
Chatamari is a traditional Newari dish that's a must-try for anyone visiting Nepal. It consists of a thin rice flour crepe topped with a variety of ingredients, such as minced meat, vegetables, eggs and spices. The end story is a savoury and flavourful pizza-like dish with a unique texture.
A Newari dish (Chatamari) also known as Nepali Pizza served on a brass plate.
Chatamari can look like many different things depending on who is making it.
The Newars are an indigenous community of the Kathmandu Valley who have their own unique cuisine that's a must-try for any food enthusiast visiting Nepal.
"Newari cuisine is renowned for its rich flavours and intricate dishes."
Chatamari is usually enjoyed as a light meal, so it's a great one to pick up in a market during a day of exploring when you discover your energy is waning.

---

--- Chunk 2 ---
9. Sel roti, a quintessential Nepali treat
Sel roti is a traditional Nepali, especially Newar, rice

To answer user queries using relevant information, we perform a retrieval step that finds the most semantically similar chunks from our FAISS index. This ensures that each LLM receives high-quality, factual context when generating its response.

For example, the query:

"What are the must-try foods for tourists visiting Kathmandu?"

retrieved the following chunk types:

 - Traditional dishes like Dal Bhat, Sel Roti, and Chatamari

 - Accommodation pricing details (peripherally relevant for tourist context)


This demonstrates that the system is able to:

 - Identify thematically appropriate chunks using MiniLM embeddings

 - Retrieve multiple valid perspectives on the same question

 - Preserve retrieval fairness by using a fixed input limit (k=4)

These retrieved chunks are then passed to different LLMs through prompt templates in the next step, enabling the generation of grounded and relevant answers.

# 3. Load All LLM Pipelines

In [9]:
# FLAN-T5 Large
flan_large_pipe = pipeline(
    "text2text-generation",
    model="google/flan-t5-large",
    tokenizer="google/flan-t5-large"
)

# FLAN-T5 Base
flan_base_pipe = pipeline(
    "text2text-generation",
    model="google/flan-t5-base",
    tokenizer="google/flan-t5-base"
)

config.json:   0%|          | 0.00/662 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/3.13G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/2.54k [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.42M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/2.20k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.40k [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/990M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/2.54k [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.42M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/2.20k [00:00<?, ?B/s]

In [10]:
# TinyLLaMA Chat
tinyllama_pipe = pipeline(
    "text-generation",
    model="TinyLlama/TinyLlama-1.1B-Chat-v1.0",
    tokenizer="TinyLlama/TinyLlama-1.1B-Chat-v1.0",
    return_full_text=False
)

config.json:   0%|          | 0.00/608 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.20G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/1.29k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.84M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/551 [00:00<?, ?B/s]

## **Final Selected Models**

| Model | Type | Why It Was Chosen |
|-------|------|-------------------|
| **FLAN-T5 Large** (`google/flan-t5-large`) | Text-to-Text | Strong instruction-following, good output structure, works reliably on Colab |
| **FLAN-T5 Base** (`google/flan-t5-base`) | Text-to-Text | Lightweight version of FLAN, included to compare how size affects performance |
| **TinyLLaMA-1.1B Chat** (`TinyLlama/TinyLlama-1.1B-Chat-v1.0`) | Chat-tuned Causal LM | Lightweight LLaMA-style chat model, gives fluent output with low memory usage |

## ***Models Attempted but Excluded***

| Model | Reason for Exclusion |
|-------|-----------------------|
| `Phi-2`, `Phi-1.5` | RAM usage too high, session crashed |
| `Mistral-7B`, `LLaMA-3` | Too large for Colab runtime, 403 errors or GPU OOM |
| Quantized GPTQ models (e.g. Falcon, Mistral via `auto_gptq`) | Required non-Colab-safe installations and crashed when loaded |
| `FLAN-UL2`, `LLaMA-13B` | Exceeded token context or failed to load due to size |

# 4. Define RAG Prompt Generator (3-Model Version)

Each language model in this project requires a slightly different prompt format, depending on its architecture and training style. To ensure fairness and compatibility, we created a `generate_rag_response()` function that:

- Embeds the query using the same MiniLM encoder used for the knowledge base
- Retrieves the top 4 most relevant text chunks from the FAISS index
- Constructs a model-specific prompt based on the retrieved context
- Generates an answer using the chosen LLM pipeline

### Prompt Format per Model

| Model           | Prompt Style                                 | Why |
|-----------------|-----------------------------------------------|-----|
| **FLAN-T5 (Large & Base)** | Uses a plain QA-style format with a `Context:` section followed by the query and `Answer:` cue | Prevents FLAN from misinterpreting the prompt as a multiple-choice test |
| **TinyLLaMA**   | Uses a role-based chat format starting with "You are a Nepal travel expert..." | Matches its instruction/chat-tuned training format |

Additionally, we:
- Limit the full input to a **maximum of 512 tokens** for FLAN models
-  Apply token-based truncation using `AutoTokenizer`
- Limit raw context to ~1400 characters for safety
- Use `return_full_text=False` in the TinyLLaMA pipeline to avoid repeating the entire prompt in the output

This prompt structure ensures that each model is fairly and effectively guided to generate helpful, relevant answers grounded in the retrieved text.

In [11]:
# Load once outside the function to avoid reloading every time
flan_tokenizer = AutoTokenizer.from_pretrained("google/flan-t5-large")

In [15]:
def generate_rag_response(query, embedder, index, chunks, generator, model_type="flan", k=4, max_chars=1800):
    # Step 1: Retrieve top-k chunks
    top_chunks = search_faiss_index(query, embedder, index, chunks, k)

    # Limit context to avoid token overflow
    context = "\n".join(top_chunks)

    if model_type.startswith("flan"):
        context = context[:1400]
        prompt = f"""Context:
{context}

{query}

Answer:"""

        inputs = flan_tokenizer(prompt, return_tensors="pt", truncation=True, max_length=512)
        prompt_trimmed = flan_tokenizer.decode(inputs["input_ids"][0], skip_special_tokens=True)

        result = generator(prompt_trimmed.strip(), max_new_tokens=300)

    elif model_type == "tinyllama":
        prompt = f"""You are a Nepal travel expert. Use the context below to answer the question briefly and clearly.
Context:
{context}

Question:
{query}

Answer:"""
        result = generator(prompt.strip(), max_new_tokens=150)

    else:
        raise ValueError(f"Unsupported model_type: {model_type}")

    return result[0]['generated_text'].strip()

#5. Load 12 Domain Specific Evaluation Questions

In [16]:
rag_questions = [
    "What is the must-try traditional foods that tourists should eat while visiting Nepal?",
    "How much do hotels typically cost in Kathmandu?",
    "What is the visa process and fee for U.S. citizens traveling to Nepal?",
    "What is the average roundtrip flight cost from the United States to Kathmandu?",
    "Which major festivals should travelers plan around when visiting Nepal?",
    "What are the best local transportation options for getting around Nepal as a tourist?",
    "What is the recommended 10-day itinerary for trekking to Everest Base Camp?",
    "What are the top 3 must-visit destinations in Nepal for first-time travelers?",
    "Which are the top 3 most popular trekking routes in Nepal?",
    "Which treks in Nepal are best for beginners?",
    "Which treks are considered the most difficult for experienced trekkers?",
    "What should tourists pack when preparing for a trek in the Himalayas?",
    "What is the cheapest hotel or accommodation tourists can find in Nepal?"
]

# 6. Run RAG Inference Loop Across 13 Questions across 3 Models

In [17]:
# Create an empty list of dicts for each model
flan_large_answers = []
flan_base_answers = []
tinyllama_answers = []

In [18]:
for i, question in enumerate(rag_questions, 1):
    print(f"FLAN-T5 Large – Q{i}: {question}")
    answer = generate_rag_response(
        query=question,
        embedder=embedder,
        index=index,
        chunks=chunks,
        generator=flan_large_pipe,
        model_type="flan"
    )
    flan_large_answers.append(answer)

FLAN-T5 Large – Q1: What is the must-try traditional foods that tourists should eat while visiting Nepal?
FLAN-T5 Large – Q2: How much do hotels typically cost in Kathmandu?
FLAN-T5 Large – Q3: What is the visa process and fee for U.S. citizens traveling to Nepal?
FLAN-T5 Large – Q4: What is the average roundtrip flight cost from the United States to Kathmandu?
FLAN-T5 Large – Q5: Which major festivals should travelers plan around when visiting Nepal?
FLAN-T5 Large – Q6: What are the best local transportation options for getting around Nepal as a tourist?
FLAN-T5 Large – Q7: What is the recommended 10-day itinerary for trekking to Everest Base Camp?
FLAN-T5 Large – Q8: What are the top 3 must-visit destinations in Nepal for first-time travelers?
FLAN-T5 Large – Q9: Which are the top 3 most popular trekking routes in Nepal?
FLAN-T5 Large – Q10: Which treks in Nepal are best for beginners?
FLAN-T5 Large – Q11: Which treks are considered the most difficult for experienced trekkers?
FLAN-T

In [19]:
for i, question in enumerate(rag_questions, 1):
    print(f"FLAN-T5 Base – Q{i}: {question}")
    answer = generate_rag_response(
        query=question,
        embedder=embedder,
        index=index,
        chunks=chunks,
        generator=flan_base_pipe,
        model_type="flan"
    )
    flan_base_answers.append(answer)

FLAN-T5 Base – Q1: What is the must-try traditional foods that tourists should eat while visiting Nepal?
FLAN-T5 Base – Q2: How much do hotels typically cost in Kathmandu?
FLAN-T5 Base – Q3: What is the visa process and fee for U.S. citizens traveling to Nepal?
FLAN-T5 Base – Q4: What is the average roundtrip flight cost from the United States to Kathmandu?
FLAN-T5 Base – Q5: Which major festivals should travelers plan around when visiting Nepal?
FLAN-T5 Base – Q6: What are the best local transportation options for getting around Nepal as a tourist?
FLAN-T5 Base – Q7: What is the recommended 10-day itinerary for trekking to Everest Base Camp?
FLAN-T5 Base – Q8: What are the top 3 must-visit destinations in Nepal for first-time travelers?
FLAN-T5 Base – Q9: Which are the top 3 most popular trekking routes in Nepal?
FLAN-T5 Base – Q10: Which treks in Nepal are best for beginners?
FLAN-T5 Base – Q11: Which treks are considered the most difficult for experienced trekkers?
FLAN-T5 Base – Q1

In [20]:
for i, question in enumerate(rag_questions, 1):
    print(f"TinyLLaMA – Q{i}: {question}")
    answer = generate_rag_response(
        query=question,
        embedder=embedder,
        index=index,
        chunks=chunks,
        generator=tinyllama_pipe,
        model_type="tinyllama"
    )
    tinyllama_answers.append(answer)

TinyLLaMA – Q1: What is the must-try traditional foods that tourists should eat while visiting Nepal?
TinyLLaMA – Q2: How much do hotels typically cost in Kathmandu?
TinyLLaMA – Q3: What is the visa process and fee for U.S. citizens traveling to Nepal?
TinyLLaMA – Q4: What is the average roundtrip flight cost from the United States to Kathmandu?
TinyLLaMA – Q5: Which major festivals should travelers plan around when visiting Nepal?
TinyLLaMA – Q6: What are the best local transportation options for getting around Nepal as a tourist?
TinyLLaMA – Q7: What is the recommended 10-day itinerary for trekking to Everest Base Camp?
TinyLLaMA – Q8: What are the top 3 must-visit destinations in Nepal for first-time travelers?
TinyLLaMA – Q9: Which are the top 3 most popular trekking routes in Nepal?
TinyLLaMA – Q10: Which treks in Nepal are best for beginners?
TinyLLaMA – Q11: Which treks are considered the most difficult for experienced trekkers?
TinyLLaMA – Q12: What should tourists pack when pr

In [21]:
print("Questions:", len(rag_questions))
print("FLAN-T5 Large:", len(flan_large_answers))
print("FLAN-T5 Base:", len(flan_base_answers))
print("TinyLLaMA:", len(tinyllama_answers))


Questions: 13
FLAN-T5 Large: 13
FLAN-T5 Base: 13
TinyLLaMA: 13


In [22]:
df_rag = pd.DataFrame({
    "Question": rag_questions,
    "FLAN-T5 Large": flan_large_answers,
    "FLAN-T5 Base": flan_base_answers,
    "TinyLLaMA": tinyllama_answers
})

In [23]:
pd.set_option("display.max_colwidth", None)
display(df_rag)

Question  \
0   What is the must-try traditional foods that tourists should eat while visiting Nepal?   
1                                         How much do hotels typically cost in Kathmandu?   
2                  What is the visa process and fee for U.S. citizens traveling to Nepal?   
3          What is the average roundtrip flight cost from the United States to Kathmandu?   
4                 Which major festivals should travelers plan around when visiting Nepal?   
5   What are the best local transportation options for getting around Nepal as a tourist?   
6             What is the recommended 10-day itinerary for trekking to Everest Base Camp?   
7           What are the top 3 must-visit destinations in Nepal for first-time travelers?   
8                              Which are the top 3 most popular trekking routes in Nepal?   
9                                            Which treks in Nepal are best for beginners?   
10                Which treks are considered the most difficult for experienced trekkers?   
11                  What should tourists pack when preparing for a trek in the Himalayas?   
12                What is the cheapest hotel or accommodation tourists can find in Nepal?   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     FLAN-T5 Large  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        Chatamari   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Average cost per night in Kathmandu is around $80.   
2   Visa fees are payable in U.S. dollars. Money exchanges and ATM services are available at the airport. Credit cards may also be accepted at the airport. Other Visa Categories: As of February 2025, foreign travelers must arrive in Nepal on a tourist visa and can only convert their visa to a different category after arrival. Check with the Department of Immigration for visa details and the on l arrive by land, you must obtain a visa. Nepali authorities do not stop all foreign citizens who enter by land to process immigration and visa documentation. If you do not have a visa and do not receive

In [24]:
df_rag.to_csv("RAG_3model_comparison_13q.csv", index=False)
print("Saved: RAG_3model_comparison_13q.csv")

Saved: RAG_3model_comparison_13q.csv


In [26]:
# -------------------------------
#  Auto-Scoring for RAG Responses
# -------------------------------

# Define scoring function based on keywords and length
def auto_score(answer, expected_keywords):
    if not isinstance(answer, str) or len(answer.strip()) == 0:
        return 1
    answer_lower = answer.lower()
    match_count = sum(1 for kw in expected_keywords if kw in answer_lower)

    if match_count >= 4:
        return 5 if len(answer.split()) > 50 else 4
    elif match_count >= 2:
        return 3
    elif match_count == 1:
        return 2
    else:
        return 1

# Define expected keywords for each of the 13 evaluation questions
question_keywords = {
    0: ["dal bhat", "momo", "sel roti", "chatamari", "thukpa"],
    1: ["hotel", "cost", "price", "kathmandu", "$", "budget", "mid-range", "luxury"],
    2: ["visa", "fee", "us citizen", "entry", "passport", "arrival", "$"],
    3: ["flight", "roundtrip", "cost", "usa", "kathmandu", "$", "airfare"],
    4: ["festival", "dashain", "tihar", "gai jatra", "bikram sambat","Holi","Mahashivaratri","Maghe Sankranti"],
    5: ["transportation", "bus", "taxi", "walk", "air", "vehicle", "car", "rickshaw"],
    6: ["kathmandu", "everest base camp", "Lukla", "Phakding", "Dingboche", "namche","bazaar", "Tengboche","Lobuche"],
    7: ["kathmandu", "pokhara", "chitwan", "Langtang", "mustang", "annapurna", "everest"],
    8: ["Mustang ", "everest", "annapurna", "langtang", "Ghorepani", "Kathmandu","Gokyo","Sherpa","Manaslu","Dhaulagiri","Kanchenjunga","Khumbu","Dolpo","safari","chitwan"],
    9: ["beginner", "easy", "short", "poon hill", "ghorepani", "langtang", "trek"],
    10: ["difficult", "experienced", "everest", "manaslu", "challenging", "technical", "kanchenjunga"],
    11: ["pack", "gear", "clothing", "boots", "jacket", "layers", "water", "trekking"],
    12: ["cheapest", "hotel", "accommodation", "budget", "$", "guest house", "hostel"]
}

# Apply scoring per model
for model_col, score_col in [
    ("FLAN-T5 Large", "FLAN Large Score"),
    ("FLAN-T5 Base", "FLAN Base Score"),
    ("TinyLLaMA", "TinyLLaMA Score")
]:
    df_rag[score_col] = df_rag.apply(
        lambda row: auto_score(row[model_col], question_keywords.get(row.name, [])),
        axis=1
    )

# Preview updated table
df_rag[["Question", "FLAN Large Score", "FLAN Base Score", "TinyLLaMA Score"]].head()


Question  \
0  What is the must-try traditional foods that tourists should eat while visiting Nepal?   
1                                        How much do hotels typically cost in Kathmandu?   
2                 What is the visa process and fee for U.S. citizens traveling to Nepal?   
3         What is the average roundtrip flight cost from the United States to Kathmandu?   
4                Which major festivals should travelers plan around when visiting Nepal?   

   FLAN Large Score  FLAN Base Score  TinyLLaMA Score  
0                 2                2                2  
1                 3                2                3  
2                 5                3                5  
3                 2                2                5  
4                 2                2                3

In [27]:
# Compute average score per model
avg_scores = {
    "FLAN-T5 Large": df_rag["FLAN Large Score"].mean(),
    "FLAN-T5 Base": df_rag["FLAN Base Score"].mean(),
    "TinyLLaMA": df_rag["TinyLLaMA Score"].mean()
}

# Print results
for model, avg in avg_scores.items():
    print(f"{model} average score: {avg:.2f} / 5")

FLAN-T5 Large average score: 2.23 / 5
FLAN-T5 Base average score: 1.85 / 5
TinyLLaMA average score: 3.77 / 5


## RAG Model Scoring Summary

## **Why I chose this Scoring Approach?**

To objectively compare LLM outputs, I used a keyword-matching heuristic that scores each model response based on:

- Relevance to the question
- Presence of expected domain keywords
- Answer length as a proxy for completeness

This approach ensures that models are evaluated on factual alignment, not just fluency or verbosity. The scoring rubric ranges from 1 (off-topic) to 5 (fully relevant and rich).

**Scoring Rubric (1-5)**

| Score | Description                                       |
| ----- | ------------------------------------------------- |
| **5** | Highly relevant, detailed, 4+ keywords, >50 words |
| **4** | Good coverage, 4+ keywords, shorter               |
| **3** | Partial relevance, 2-3 keywords                   |
| **2** | Minimal match, 1 keyword                          |
| **1** | Off-topic, incorrect, or missing content          |



**Average Scores (13 Questions)**

| Model             | Avg. Score (out of 5) | Observations                                                               |
| ----------------- | --------------------- | -------------------------------------------------------------------------- |
| **TinyLLaMA**     | **3.77**              | High relevance, strong reasoning, but verbose and hallucination-prone      |
| **FLAN-T5 Large** | **2.23**              | Decent factual answers after prompt fix, but often too short or incomplete |
| **FLAN-T5 Base**  | **1.85**              | Very brief answers, limited reasoning, rarely complete                     |



## Qualitative Analysis: Hallucination, Reasoning & Patterns
Beyond relevance scoring, I conducted a qualitative evaluation of model responses to identify deeper behavioral traits such as hallucination, reasoning strength, fluency, and response patterns.

 ## Observed Patterns & Behaviors

| Model             | Strengths                                                                     | Weaknesses                                                                  | Hallucination Behavior                                                  |
| ----------------- | ----------------------------------------------------------------------------- | --------------------------------------------------------------------------- | ----------------------------------------------------------------------- |
| **TinyLLaMA**     |  Very detailed responses<br> Strong reasoning<br> Fluent writing style     |  Overgeneration<br> Mixed irrelevant topics<br> Needs token control      |  Moderate- High: Hallucinated some festivals, foods, and trek names    |
| **FLAN-T5 Large** |  Better after prompt fix<br>Factually grounded answers<br> Good structure |  Incomplete at times<br> Sometimes too vague<br> Initially failed prompt |  Low - Moderate: Sometimes off-topic (e.g. wrong domain chunk) |
| **FLAN-T5 Base**  |  Stable, short responses<br> Fast inference                                 |  Very brief<br> Minimal explanation or reasoning<br> Often too generic   |  Low: Rare hallucination, but lacked meaningful depth                  |


## Reasoning & Output Style Summary

| Model             | Reasoning   | Fluency  | Output Style           | Needs Improvement On...            |
| ----------------- | ----------- | -------- | ---------------------- | ---------------------------------- |
| **TinyLLaMA**     |  High     | High  | Descriptive, verbose   | Token length, hallucination filter |
| **FLAN-T5 Large** |  Moderate |  Clear  | Factual, but short     | Prompt tuning, answer completeness |
| **FLAN-T5 Base**  | Low      | Stable | One-liners, minimalist | Depth, context synthesis           |
